tests/MongoDB.Driver.Core.Tests/Core/Authentication/MongoAWSAuthenticatorTests.cs

```diff
@@ -0,0 +1,188 @@
+* limitations under the License.
+*/
+
+using System;
+using System.Net;
+using System.Threading;
+using FluentAssertions;
+using MongoDB.Bson;
+using MongoDB.Driver.Core.Clusters;
+using MongoDB.Driver.Core.Servers;
+using MongoDB.Driver.Core.Helpers;
+using MongoDB.Driver.Core.Misc;
+using MongoDB.Driver.Core.WireProtocol.Messages;
+using Xunit;
+using MongoDB.Driver.Core.Connections;
+using MongoDB.Bson.TestHelpers.XunitExtensions;
+using MongoDB.Driver.Core.Authentication;
+using Moq;
+using MongoDB.Bson.IO;
+
+namespace MongoDB.Driver.Core.Tests.Core.Authentication
+{
+    public class MongoAWSAuthenticatorTests
+    {
+        // private constants
+        private const int ClientNonceLength = 32;
+
+        // private static
+        private static readonly IRandomByteGenerator __randomByteGenerator = new DefaultRandomByteGenerator();
+
+        [Theory]
+        [ParameterAttributeData]
+        public void AuthenticateShouldWorkAsExpected(
+            [Values(false, true)] bool async)
+        {
+            var dateTime = DateTime.UtcNow;
+
+            var serverId = new ServerId(new ClusterId(), new DnsEndPoint("localhost", 27017));
+            var connectionDescription = new ConnectionDescription(
+                new ConnectionId(serverId),
+                new IsMasterResult(new BsonDocument("ok", 1).Add("ismaster", 1)),
+                new BuildInfoResult(new BsonDocument("version", "4.0.0")));
+
+            var clientNonce = __randomByteGenerator.Generate(ClientNonceLength);
+            var serverNonce = Combine(clientNonce, __randomByteGenerator.Generate(ClientNonceLength));
+            var host = "sts.amazonaws.com";
+            var credential = new UsernamePasswordCredential("$external", "permanentuser", "FAKEFAKEFAKEFAKEFAKEfakefakefakefakefake");
+            
+            AwsSignatureVersion4.SignRequest(
+                dateTime,
+                credential.Username,
+                credential.GetInsecurePassword(),
+                null,
+                serverNonce,
+                host,
+                out var authHeader,
+                out var timestamp);
+
+            var dateTimeProviderMock = new Mock<IClock>();
+            dateTimeProviderMock.Setup(x => x.UtcNow).Returns(dateTime);
+
+            var randomByteGeneratorMock = new Mock<IRandomByteGenerator>();
+            randomByteGeneratorMock.Setup(x => x.Generate(It.IsAny<int>())).Returns(clientNonce);
+
+            var expectedClientFirstMessage = new BsonDocument()
+                .Add("r", new BsonBinaryData(clientNonce))
+                .Add("p", new BsonInt32('n'));
+            var expectedClientSecondMessage = new BsonDocument()
+                .Add("a", authHeader)
+                .Add("d", timestamp);
+
+            var serverFirstMessage = new BsonDocument()
+                .Add("s", new BsonBinaryData(serverNonce))
+                .Add("h", new BsonString(host));
```

Comment:
The above three variables would more idiomatically be initialized using:

    var expectedClientFirstMessage = new BsonDocument
    {
        { "r", clientNonce },
        { "p", (int)'n' }
    };
    var expectedClientSecondMessage = new BsonDocument()
    {
        { "a", authHeader },
        { "d", timestamp }
    };

    var serverFirstMessage = new BsonDocument
    {
        { "s", serverNonce },
        { "h", host }
    };